File: source/python/url_shortener.py
url_to_short = {}
short_to_url = {}
counter = 1000

def encode_id(n):
    chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
    result = ""
    while n > 0:
        rem = n % 62
        result = chars[rem] + result
        n = (n - rem) // 62
    return result

def shorten(url):
    global counter
    if url in url_to_short:
        return url_to_short[url]
    counter += 1
    code = encode_id(counter)
    short = f"https://sho.rt/{code}"
    url_to_short[url] = short
    short_to_url[short] = url
    return short

def resolve(short_url):
    return short_to_url.get(short_url)

urls = [
    "https://example.com/long/path",
    "https://docs.hemlock.org/stdlib",
    "https://github.com/hemlang/42",
    "https://example.com/long/path",
]

print("=== URL Shortener ===")
for url in urls:
    short = shorten(url)
    print(f"  {url}")
    print(f"  -> {short}")
    print("")

print("=== Resolve Short URLs ===")
test_short = shorten("https://docs.hemlock.org/stdlib")
resolved = resolve(test_short)
print(f"  {test_short} -> {resolved}")

s1 = shorten("https://example.com/long/path")
s2 = shorten("https://example.com/long/path")
if s1 == s2:
    print("\nDuplicate URL returns same short URL: ok")

File: source/javascript/url_shortener.js
// URL shortener using in-memory maps
const urlToShort = {};
const shortToUrl = {};
let counter = 1000;

function encodeId(id) {
    const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    let result = "";
    let n = id;
    while (n > 0) {
        const rem = n % 62;
        result = chars[rem] + result;
        n = Math.floor((n - rem) / 62);
    }
    return result;
}

function shorten(url) {
    if (urlToShort[url] != null) return urlToShort[url];
    counter++;
    const code = encodeId(counter);
    const short = `https://sho.rt/${code}`;
    urlToShort[url] = short;
    shortToUrl[short] = url;
    return short;
}

function resolve(shortUrl) {
    return shortToUrl[shortUrl] != null ? shortToUrl[shortUrl] : null;
}

const urls = [
    "https://example.com/long/path",
    "https://docs.hemlock.org/stdlib",
    "https://github.com/hemlang/42",
    "https://example.com/long/path"
];

console.log("=== URL Shortener ===");
for (const url of urls) {
    const short = shorten(url);
    console.log(`  ${url}`);
    console.log(`  -> ${short}`);
    console.log("");
}

console.log("=== Resolve Short URLs ===");
const testShort = shorten("https://docs.hemlock.org/stdlib");
const resolved = resolve(testShort);
console.log(`  ${testShort} -> ${resolved}`);

const s1 = shorten("https://example.com/long/path");
const s2 = shorten("https://example.com/long/path");
if (s1 === s2) console.log("\nDuplicate URL returns same short URL: ok");
